pkg/endpoint/policy.go

```diff
@@ -163,81 +148,151 @@ func (e *Endpoint) setNextPolicyRevision(revision uint64) {
 // however, and it is possible that policy update succeeds for some endpoints,
 // while it fails for other endpoints.
 //
-// Returns:
-//   - err: any error in obtaining information for computing policy, or if
+// Failure may be due to any error in obtaining information for computing policy,
+// or if policy could not be generated given the current set of rules in the repository.
 //
-// policy could not be generated given the current set of rules in the
-// repository.
-// Must be called with endpoint mutex held.
-func (e *Endpoint) regeneratePolicy() (retErr error) {
-	var forceRegeneration bool
+// endpoint lock must NOT be held. This is because the ipcache needs to be able to
+// make progress while generating policy, and *that* needs the endpoint unlocked to call
+// ep.ApplyPolicyMapChanges. Specifically, computing policy may cause identity allocation
+// which requires ipcache progress.
+//
+// buildMutex MUST be held, and not released until setDesiredPolicy and
+// updateRealizedState have been called
+//
+// There are a few fields that depend on this exact configuration of locking:
+//   - ep.desiredPolicy: ep.mutex must be locked between writing this and committing to
+//     the policy maps, or else policy drops may occur
+//   - ep.policyRevision: ep.mutex and ep.buildMutex must be held to write to this, or else
+//     the policyRevision may go backwards from no-op bumps from policy ingestion
```

Comment:
Hmm, interesting. I'll do some more digging. I *definitely* saw consistent test failures around endpoints failing to reach a certain policy revision, and the solution was *more locking*, which 100% eliminated that failure.

Reply:
Well, I can't reproduce the failure now, even after removing the lock. I'll update the inaccurate comment. There must have been some other incorrect logic also cleaned up.

Reply:
@joestringer this brings up another question: how much can / should we rely on exported methods only being called by the event queue?

Reply:
Yeah good point, ideally the methods would be private so there's no other path. But at the same time, I think that if we don't reinforce that assumption, then we open a whole can of worms around the style of the Endpoint internals because we have to know how the entire rest of the daemon may end up calling into Endpoint. This thread is one such example, there are surprising interactions between the way policy ingestion and internal endpoint policy calculation works.